### PR TITLE
API Conventions for Late-initializers

### PR DESCRIPTION
**Pull request by erictune**
Formalizing what @bgrant0607 and I discussed in the context of the SecurityContext changes.

### CODE REVIEW

File: docs/api-conventions.md [hunk 1]
```diff
@@ -294,6 +294,27 @@ Incorporating the default values into the `Spec` ensures that `Spec` depicts the
...
+-------------------
+Late initialization is when resource fields are set by a system controller
+after an object is created/updated.
+
+For example, the scheduler sets the pod.spec.host field after the pod is created.
+
+Late-initializers should only make the following types of modifications:
+  - Setting previously unset fields
+  - Adding keys to maps
+  - Adding values to arrays which have mergeable semantics (mergePatch attribute in go definition of type).
```
Comment (pmorie) [inline docs/api-conventions.md]: Nit: newline?

Comment (bgrant0607) [inline docs/api-conventions.md]: `patchStrategy:"merge"`

File: docs/api-conventions.md [hunk 2]
```diff
@@ -294,6 +294,27 @@ Incorporating the default values into the `Spec` ensures that `Spec` depicts the
...
+These conventions allow a user (with sufficient privilege) to override any system-default behaviors by setting
+the fields that would otherwise have been defaulted. 
+They also allow the component which does the defaulting to use strategic merge patch, which facilitates composition and
```
Comment (bgrant0607) [inline docs/api-conventions.md]: More importantly, it enables updates from users to be merged with changes made during late initialization, using strategic merge patch, as opposed to clobbering the change.

File: docs/api-conventions.md (hunk 1 above)
Comment (erictune) [inline docs/api-conventions.md]: both fixed.

File: docs/api-conventions.md (hunk 2 above)
Comment (erictune) [inline docs/api-conventions.md]: fixed.

**Comment by bgrant0607** [comment]
LGTM. Thanks much.

**Comment by derekwaynecarr** [comment]
This is helpful to lay out.  Agree with the admission control comments for those handlers that modify incoming objects.

**Pull request merged**